### PR TITLE
Tile stream

### PR DESCRIPTION
**Pull request by at88mph**
Sorry, it's a big one.

The `StreamingImageData` class will write out properly, and delegate reading into it to the Tiler.  The Tiler had to be modified to allow proper streaming from the `RandomAccess` Object.

The way the `StreamingImageData` is used can be a little cumbersome given the amount of state the `Fits` Object, and its HDUs, hold.  I've had to clone and make copies where necessary to remove references.

With these changes, one can set the `streamWriteFlag` in the `Fits` Object to get a `StreamingImageData` instance in the `ImageHDU`s.  When iterating over HDUs, however, and writing out to another file for tiling, one must still create a new `StreamingImageData` Object but clone the `StandardImageTiler` to preserve the underlying `ReadAccess` object.

### CODE REVIEW

**Comment by at88mph** [comment]
There was effort put into preserving the Maven build, as well.

File: src/main/java/nom/tam/image/ImageTiler.java [hunk 1]
```diff
@@ -46,6 +49,8 @@
 
     Object getTile(int[] corners, int[] lengths) throws IOException;
 
+    void getTile(ArrayDataOutput output, int[] corners, int[] lengths) throws FitsException, IOException;
```
Comment (pdowler) [inline src/main/java/nom/tam/image/ImageTiler.java:52]: I was not expecting to a need for the above method. Doesn't subsection (corners and lengths) happen during read and therefore everything that is read gets written out?

File: src/test/java/nom/tam/util/BufferedFileTest.java [hunk 2]
```diff
@@ -135,4 +138,116 @@ public void testReadFully2() throws IOException {
...
+            input.read(resultData);
+            Assert.assertEquals("Wrong data length.", expectedData.length, resultData.length);
+
+            for (int i = 0; i < resultData.length; i++) {
+                Assert.assertTrue("Wrong value at " + i, resultData[i]);
+            }
+        } finally {
+            file.close();
+        }
+    }
```
Comment (pdowler) [inline src/test/java/nom/tam/util/BufferedFileTest.java:252]: Mainly: it seems unclear what these new test methods are trying to test w.r.t the BufferedFile class...

Secondarily: It's worth noting that many FITS data arrays are stored as integers (BITPIX=16 means short, BITPIX=32 means int, could be that BITPIX=64 is used as well but haven't seen it. Testing byte, float, and boolean seems incomplete to me and further reinforces the lack of clarity for the main issue above.

File: pom.xml [hunk 3]
```diff
@@ -6,8 +6,8 @@
 		<artifactId>oss-parent</artifactId>
 		<version>7</version>
 	</parent>
-	<groupId>gov.nasa.gsfc.heasarc</groupId>
-	<version>1.15.3-pdowler</version>
+	<groupId>org.opencadc</groupId>
+	<version>1.17.0</version>
```
Comment (pdowler) [inline pom.xml:10]: Would like to see version 1.16.1 here and in build.gradle

**Review by pdowler — CHANGES_REQUESTED** [review]

File: src/main/java/nom/tam/util/MemoryUsage.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        final Runtime runtime = Runtime.getRuntime();
+
+        // Run the garbage collector
+        final long start = System.currentTimeMillis();
+        runtime.gc();
+        System.out.println("Garbage Collection: OK (" + (System.currentTimeMillis() - start) + "ms)");
+
+        // Calculate the used memory
+        final long memory = runtime.totalMemory() - runtime.freeMemory();
+
```
Comment (pdowler) [inline src/main/java/nom/tam/util/MemoryUsage.java:128]: The output below should use the standard logging rather than System.out ... debug level.

File: src/test/java/nom/tam/util/BufferedFileTest.java (hunk 2 above)
Comment (at88mph) [reply src/test/java/nom/tam/util/BufferedFileTest.java:252]: OK.
The `StandardImageTiler` on line 317 makes use of the `MemoryUsage` class.  It just checkpoints every 200 skips along the first dimension of the array.  That tiler will move along the first dimension (using the `incrementPosition` method and then perform a seek in the file.  The `MemoryUsage` class will checkpoint every nth loop to print out, but only after the call to `read(data output, ...)` has been called to show that nothing beyond the cached headers (ones already read) are being put into memory.

File: src/main/java/nom/tam/util/MemoryUsage.java (hunk 4 above)
Comment (at88mph) [reply src/main/java/nom/tam/util/MemoryUsage.java:135]: Done.  This library is a mix of Java Logging and System.out.println.

**Review by pdowler — COMMENTED** [review]
This PR is still a _draft_ and to be honest the amount of "wiring" needed to do this is hard to follow so I can't actually tell if the work is complete or correct... As noted, part of the work still has to be done by the client to make streaming read-to-write work.

File: src/test/java/nom/tam/util/BufferedFileTest.java (hunk 2 above)
Comment (pdowler) [reply src/test/java/nom/tam/util/BufferedFileTest.java:252]: Ah, didn't spot that. thnx. Might be good for the MemoryUsage.checkpoint to check the log level and do nothing if it isn't actually going to output any info. I don't think it's good practice to call `System.gc()` in an app server (esp. frequently and/or in a loop)

Comment (at88mph) [reply src/test/java/nom/tam/util/BufferedFileTest.java:252]: Adding that test is a good suggestion and worked well.  It also exposed that I missed testing with `char`.

**Review by pdowler — APPROVED** [review]
